### PR TITLE
SELV3-575: Added functionality for fetching DHIS2 API datasets.

### PR DESCRIPTION
**Pull request by sradziszewski**
Added DhisDataService
Fixed MessageKeys class duplication.
Added RestTemplate configuration.
Added model classes for DHIS2 API dataset.
Added helper classes: RequestHelper and RequestParameters.
JIRA task link: [SELV3-575](https://openlmis.atlassian.net/browse/SELV3-575)

[SELV3-575]: https://openlmis.atlassian.net/browse/SELV3-575?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/org/openlmis/integration/dhis2/dto/dhis/DataElement.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+package org.openlmis.integration.dhis2.dto.dhis;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
+public class DataElement {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/integration/dhis2/dto/dhis/DataElement.java:25]: I think that a knowledge about DHIS2 data model might be lost after few months. 
Please add some JavaDoc which roughly explains what the class is about - like DataElement is a description of single data value, not a calculated value itself but 'a type/kind' (?).
Think what you would like to read here after a year you code it.

File: src/main/java/org/openlmis/integration/dhis2/dto/dhis/DhisDataset.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Setter;
+import lombok.ToString;
+
+/**
+ * Objects of this class represent data retrieved from the DHIS2 API.
+ */
+@Getter
+@Setter
+@ToString
+public class DhisDataset {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/integration/dhis2/dto/dhis/DhisDataset.java:29]: (same as DataElement)
I think that a knowledge about DHIS2 data model might be lost after few months. 
Please add some JavaDoc which roughly explains what the class is about - like DataElement is a description of single data value, not a calculated value itself but 'a type/kind' (?).
Think what you would like to read here after a year you code it.

File: src/main/java/org/openlmis/integration/dhis2/dto/dhis/OrganisationUnit.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+package org.openlmis.integration.dhis2.dto.dhis;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
+public class OrganisationUnit {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/integration/dhis2/dto/dhis/OrganisationUnit.java:25]: (same as DataElement)
I think that a knowledge about DHIS2 data model might be lost after few months. 
Please add some JavaDoc which roughly explains what the class is about - like DataElement is a description of single data value, not a calculated value itself but 'a type/kind' (?).
Think what you would like to read here after a year you code it.

**Review by pwargulak — CHANGES_REQUESTED** [review]
Small JavaDoc and should be fine.

File: src/main/java/org/openlmis/integration/dhis2/util/RequestParameters.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+package org.openlmis.integration.dhis2.util;
+
+import com.google.common.collect.Maps;
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class RequestParameters {
+  private final Map<String, Object> params = Maps.newHashMap();
+
```
[the remark this answers is not in the document]
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/util/RequestParameters.java:25]: Empty line added.

File: src/main/java/org/openlmis/integration/dhis2/dto/dhis/DhisDataset.java (hunk 2 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/dto/dhis/DhisDataset.java:33]: I added JavaDoc.

File: src/main/java/org/openlmis/integration/dhis2/dto/dhis/OrganisationUnit.java (hunk 3 above)
Comment (sradziszewski) [reply src/main/java/org/openlmis/integration/dhis2/dto/dhis/OrganisationUnit.java:32]: I added JavaDoc.

File: src/test/java/org/openlmis/integration/dhis2/util/RequestHelperTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  public void shouldCreateEntity() {
+    String token = "r4nd0m70k3n";
+    String tokenPrefix = "Prefix";
+
+    HttpEntity entity = RequestHelper.createEntity(token, tokenPrefix);
+
+    assertThat(entity.getHeaders().get(HttpHeaders.AUTHORIZATION),
+            is(singletonList("Prefix r4nd0m70k3n")));
+  }
+}
```
[the remark this answers is not in the document]
Comment (sradziszewski) [reply src/test/java/org/openlmis/integration/dhis2/util/RequestHelperTest.java:61]: Empty line added.

File: src/test/java/org/openlmis/integration/dhis2/util/RequestParametersTest.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    assertThat(toMap(parameters), not(hasKey("a")));
+  }
+
+  private Map<String, Object> toMap(RequestParameters parameters) {
+    Map<String, Object> map = Maps.newHashMap();
+    parameters.forEach(e -> map.put(e.getKey(), e.getValue()));
+
+    return map;
+  }
+}
```
[the remark this answers is not in the document]
Comment (sradziszewski) [reply src/test/java/org/openlmis/integration/dhis2/util/RequestParametersTest.java:48]: Empty line added.

**Review by anawrotsoldevelo — APPROVED** [review]

**Comment by sradziszewski** [comment]
@pwargulak 
After consulting with @anawrotsoldevelo, I added the **code** field to the Organization Unit model.
I also changed the type of the dataSetElements field in the DhisDataSet class and added a getter to bypass the **empty wrapper** from the DHIS2 API. Example:
```
"dataSetElements": [
        {
            "dataElement": {
                "name": "Plague (Deaths < 5 yrs)",
                "id": "lXolhoWewYH"
            }
        },
        {
            "dataElement": {
                "name": "Yellow Fever (Deaths < 5 yrs)",
                "id": "USBq0VHSkZq"
            }
        }
]
```

**Review by pwargulak — APPROVED** [review]
Looks fine

**Pull request merged**